### PR TITLE
Feature enhancement

### PR DESCRIPTION
**Pull request by abhinavachanta**
This feature is to calculate and record the active, passive and overall timing of sender component in the simulator. 
The function is written in '\include\phase_observer\phase_observer.hpp' file
The output is stored in 'phase_output.txt' in /data folder.

Pls, review. Thanks.

### CODE REVIEW

File: .gitmodules
Comment (syedomarazam) [inline .gitmodules]: no steps to run git submodules. see task-D pull request comments

File: data/phase_output.txt
Comment (syedomarazam) [inline data/phase_output.txt]: filename is not self-explanatory. You are doing this for the sender component.  Inside file state for which component this statistics data is for

File: include/phase_observer/phase_observer.hpp
Comment (syedomarazam) [inline include/phase_observer/phase_observer.hpp]: you must create .cpp file and implement this functionality. .hpp file will contain only function definitions.

File: src/top_model/main.cpp
Comment (syedomarazam) [inline src/top_model/main.cpp]: declarations must be done at top

**Review by syedomarazam — CHANGES_REQUESTED** [review]
Task-D changes are missing for sender, subnet and receiver .cpp files.

**Review by syedomarazam — CHANGES_REQUESTED** [review]
Task-D changes are missing for sender, subnet and receiver .cpp files.

**Comment by abhinavachanta** [comment]
> Task-D changes are missing for sender, subnet and receiver .cpp files.

We implemented the task D and Task E in different branches simultaneously. Hence task D changes are not in this Task E branch. On merging everything will be available on master.

File: README.md
Comment (pcsmartapps) [inline README.md]: Please see comments in Part-D

**Review by pcsmartapps — CHANGES_REQUESTED** [review]

**Comment by pcsmartapps** [comment]
Please solve outstanding issues pointed in the previous task as they stand true here, so i can run and test the code properly.

File: README.md
Comment (pcsmartapps) [reply README.md]: To fix the problem i suggest that you use the following command (git submodule update --init
) to invoke and update the submodules instead  of the command you have in line 131. Please addresses this issue and other points to be able to review your code.

**Comment by syedomarazam** [comment]
@abhinavachanta @jrajeev459 : Please fix all the issues asap.

1) README.md file (table of content hyperlink is not working)
2) Implement convert_output, phase_observer functionalities with .hpp and .cpp file
3) git submodules is not working, we aren't able to compile your code in Task#4 and Task#5

**Comment by abhinavachanta** [comment]
Hi. I have made the changes as per the review comments. Please review and approve if everything looks fine.

File: src/top_model/main.cpp
Comment (syedomarazam) [inline src/top_model/main.cpp:231]: when I execute ABP.exe, a set of values are getting printed on the console. I don't understand why.

**Review by syedomarazam — CHANGES_REQUESTED** [review]
phase_observer

why are these values getting printed on console?

output_covert is not implemented in sender, receiver and subnet modules.

File: src/top_model/main.cpp
Comment (jrajeev459) [reply src/top_model/main.cpp:231]: i have absolutely no idea why they are printed. i don't get any printed values.

**Review by pcsmartapps — CHANGES_REQUESTED** [review]
when i ran the simulator, i enter : ./bin/src/ABP ./data/input_abp_0.txt
i get:
Sender active time = 662 ms 
Sender overall time = 540 ms
**Sender passive time = -122 ms**
why the passive time is negative. Time can not be negative !

**Review by pcsmartapps — CHANGES_REQUESTED** [review]
Why changes in D that outputs converted_receiver_output, converted_sender_output and converted_subnet_output are not included in feature enhancement (E). Should they be included or not?

**Comment by syedomarazam** [comment]
> Why changes in D that outputs converted_receiver_output, converted_sender_output and converted_subnet_output are not included in feature enhancement (E). Should they be included or not?

@pcsmartapps they were implementing both tasks in parallel so that why those changes aren’t there in task E. i agreed last night as they are going to merge task d D and Task E to master anyways

**Comment by pcsmartapps** [comment]
> > Why changes in D that outputs converted_receiver_output, converted_sender_output and converted_subnet_output are not included in feature enhancement (E). Should they be included or not?
> 
> @pcsmartapps they were implementing both tasks in parallel so that why those changes aren’t there in task E. i agreed last night as they are going to merge task d D and Task E to master anyways

@syedomarazam and @abhinavachanta , Now part D is merged to master, are converted output files will be included or not in Part E

File: src/top_model/main.cpp
Comment (abhinavachanta) [reply src/top_model/main.cpp:231]: > when I execute ABP.exe, a set of values are getting printed on the console. I don't understand why.

Hi, I have removed the unwanted cout commands which were printing values on console. Thanks.
image

File: src/top_model/main.cpp
Comment (mohamedali5) [inline src/top_model/main.cpp:228]: You could specify what part E actually is.

"Part E - Feature enhancement"

**Comment by abhinavachanta** [comment]
> > > Why changes in D that outputs converted_receiver_output, converted_sender_output and converted_subnet_output are not included in feature enhancement (E). Should they be included or not?
> > 
> > 
> > @pcsmartapps they were implementing both tasks in parallel so that why those changes aren’t there in task E. i agreed last night as they are going to merge task d D and Task E to master anyways
> 
> @syedomarazam and @abhinavachanta , Now part D is merged to master, are converted output files will be included or not in Part E

Hi, 
The part E (phase observer) is not dependent on part D test files. It is only dependent on simulator src and output files which are already present in Part E branch. 
Since part E logic and implementation is independent and does not depend on part D test files, hence we are not adding those in part E because we worked on part D and part E in parallel. All code will be available together in master when part E is merged to master. Thanks. Let me know if any concerns

**Comment by abhinavachanta** [comment]
> when i ran the simulator, i enter : ./bin/src/ABP ./data/input_abp_0.txt
> i get:
> Sender active time = 662 ms
> Sender overall time = 540 ms
> **Sender passive time = -122 ms**
> why the passive time is negative. Time can not be negative !

Hi,

Our feature (part E) considers the 'converted_output.txt' file from part D as its input to determine the phase timings.  This 'converted_output.txt' file in part D was generated using ./data/input_abp_1.txt.

I have just now added this information in readme file also to give information to the user. 
Thanks.

**Comment by pcsmartapps** [comment]
> > when i ran the simulator, i enter : ./bin/src/ABP ./data/input_abp_0.txt
> > i get:
> > Sender active time = 662 ms
> > Sender overall time = 540 ms
> > **Sender passive time = -122 ms**
> > why the passive time is negative. Time can not be negative !
> 
> Hi,
> 
> Our feature (part E) considers the 'converted_output.txt' file from part D as its input to determine the phase timings. This 'converted_output.txt' file in part D was generated using ./data/input_abp_1.txt.
> 
> I have just now added this information in readme file also to give information to the user.
> Thanks.

Hello All,
I looked at your feedback. Whether Output converted files should be of  part E or not will be decided by Professor Cristina?  I think part D should be included in E. It would be for her to make the final call. I will approve it and would not bear any responsibility if part D should be carried over and included in part E.

As for the second issue that I have regarding to negative time. It seems that you are subtracting the active value from overall value. Nevertheless, I would not know the technical details that you have agreed with Professor Cristina to do in part E. it would be under your responsibility when I approve this merge that you have implemented the enhancements correctly and up to Professor Cristina requirements.  Please let me know.

**Review by mohamedali5 — APPROVED** [review]

**Review by syedomarazam — APPROVED** [review]
Approved

**Review by pcsmartapps — APPROVED** [review]
There are some comments that i mentioned and Abhinava said that he will discuss and agree with Professor Cristina.

**Comment by cruizm** [comment]
There is no git ignore file in the project

**Pull request merged**
